docs/docs/reference/other-new-features/explicit-nulls.md

```diff
@@ -0,0 +1,296 @@
+---
+layout: doc-page
+title: "Explicit Nulls"
+---
+
+This proposal describes a modification to the Scala type system that makes reference types
+(anything that extends `AnyRef`) _non-nullable_.
+
+This means the following code will no longer typecheck:
+```
+val x: String = null // error: found `Null`,  but required `String`
+```
+
+Instead, to mark a type as nullable we use a [type union](https://dotty.epfl.ch/docs/reference/new-types/union-types.html)
+```
+val x: String|Null = null // ok
+```
+
+Explicit nulls are enabled via a `-Yexplicit-nulls` flag, so they're an opt-in feature.
+
+Read on for details.
+
+## New Type Hierarchy
+
+When explicit nulls are enabled, the type hierarchy changes so that `Null` is subtype only of
+`Any`, as opposed to every reference type.
+
+This is the new type hierarchy:
+![](../../images/explicit-nulls/explicit-nulls-type-hierarchy.png "Type Hierarchy for Explicit Nulls")
+
+After erasure, `Null` remains a subtype of all reference types (as forced by the JVM).
+
+## Unsoundness
+
+The new type system is unsound with respect to `null`. This means there are still instances where an expressions has a non-nullable type like `String`, but its value is `null`.
+
+The unsoundness happens because uninitialized fields in a class start out as `null`:
+```scala
+class C {
+  val f: String = foo(f)
+  def foo(f2: String): String = if (f2 == null) "field is null" else f2
+}
+val c = new C()
+// c.f == "field is null"
+```
+
+Enforcing sound initialization is a non-goal of this proposal. However, once we have a type
+system where nullability is explicit, we can use a sound initialization scheme like the one
+proposed by @liufengyun and @biboudis in [https://github.com/lampepfl/dotty/pull/4543](https://github.com/lampepfl/dotty/pull/4543) to eliminate this particular source of unsoundness.
+
+## Equality
+
+Because of the unsoundness, we need to allow comparisons of the form `x == null` or `x != null`
+even when `x` has a non-nullable reference type (but not a value type). This is so we have an
+"escape hatch" for when we know `x` is nullable even when the type says it shouldn't be.
```

Comment:
I am still convinced that this is a bad idea. We shouldn't expose an escape hatch to test for null under unsoundness. There is exactly 0 use case for this. If you observe an unsound `null`, you are already in the realm of buggy programs. There is no valid program that will need this.

And even if I'm wrong and someone really really needs an escape hatch for this, there is *anyway* another workaround, which is `(x: Any) == null`.

Reply:
We will revert this. The original motivation was that it would ease migration to explicit nulls, but I don't have the data to back up that assertion at the moment. We can always add things back if and when they're needed down the line.

Reply:
@odersky @sjrd where in Dotty can we find the logic that determines whether a comparison `a == b` is valid? For some reason, the following typechecks
```scala
val s: String = ???
if (s == null) {}
```
even though `s` has its `Eql` instance defined `Eql[String, String]`, but we haven't defined one for `Null`. All the logic we could find was in `Implicits.scala`, in `synthesizedEq`, but that seems to be related to finding the "fallback" instance `Eq[Any, Any]`.

Where should we restrict comparisons like the one above. Thanks.

Reply:
I have found the logic in `Implicits.scala`, and I disallow the comparison between Object and Null for explicit nulls.